### PR TITLE
Allow predicate in Cols

### PR DESCRIPTION
**Pull request by bkamins**
Fixes https://github.com/JuliaData/DataFrames.jl/issues/2875

I was thinking that we might make it even more flexible and allow predicate to be mixed with other selectors in `Cols`, but for now I proposed a restrictive version. @nalimilan - do you think we should be more flexible?

### CODE REVIEW

File: src/other/index.jl [hunk 1]
```diff
@@ -221,6 +221,17 @@ end
...
+@inline Base.getindex(x::AbstractIndex, idx::Cols{<:Tuple{Function}}) =
+    findall(idx.cols[1], names(x))
+
+"""
+    Cols(f::Function)
+
+Select the columns whose names passed to the `f` predicate as strings return `true`.
+As a special case if `:` is passed (the `Colon` function) select all columns.
+"""
+Cols
```
Comment (nalimilan) [inline src/other/index.jl:234]: Shouldn't this be documented in the `select`, etc. docstrings?

**Comment by nalimilan** [comment]
Being restrictive sounds OK for now. Why does `:` need a special support? I would have thought it could be treated like any selector that can be included in `Cols`?

**Comment by bkamins** [comment]
> Why does `:` need a special support?

Ah - you are right. `:` is not a function but a constructor so it is not a subtype of `Function` so special treatment is not needed. I will fix this.

File: src/other/index.jl (hunk 1 above)
Comment (bkamins) [reply src/other/index.jl:234]: I think we should add this docstring. Apart from this all special selectors now are only mentioned in https://github.com/JuliaData/DataFrames.jl/blob/main/docs/src/lib/indexing.md. We should decide on a good place to document them in general, especially when we will start allowing broadcasting them. Do you see any good place?

**Comment by bkamins** [comment]
Ah - I was right originally `Colon` is a constructor but it is defined as:
```
struct Colon <: Function
end
```
so it is a `Function`. My original definition thus was required as otherwise `Colon(:)` treats `:` as a predicate. Things are tough.

File: src/other/index.jl (hunk 1 above)
Comment (bkamins) [reply src/other/index.jl:234]: @nalimilan I tried to improve the documentation. The PR should be good to re-review now.

File: docs/src/lib/indexing.md [hunk 2]
```diff
@@ -26,7 +26,17 @@ The rules for a valid type of index into a column are the following:
...
+    * an `Cols` expression (see [DataAPI.jl](https://github.com/JuliaData/DataAPI.jl));
+      the `Cols(idxs...)` selects the union of the selections in `idxs`; in particular
+      `Cols()` selects no columns and `Cols(:)` selects all columns; a special rule is
+      `Cols(predicate)`, where `precicate` is a predicate function; in this case
+      the columns whose names passed to the `predicate` predicate as strings return `true`
+      are selected.
+    * `Between` expression (see [DataAPI.jl](https://github.com/JuliaData/DataAPI.jl));
+       the `Between(first, last)` selects the columns between `first` and `last`;
+    * `All` expression (see [DataAPI.jl](https://github.com/JuliaData/DataAPI.jl));
+      the `All()` selects all columns, equivalent to `:`;
```
Comment (nalimilan) [inline docs/src/lib/indexing.md:39]:
```suggestion
      the `Not(idx)` selects all indices not in the passed `idx`;
    * a `Cols` expression (see [DataAPI.jl](https://github.com/JuliaData/DataAPI.jl));
      `Cols(idxs...)` selects the union of the selections in `idxs`; in particular
      `Cols()` selects no columns and `Cols(:)` selects all columns; a special rule is
      `Cols(predicate)`, where `predicate` is a predicate function; in this case
      the columns whose names passed to `predicate` as strings return `true`
      are selected.
    * a `Between` expression (see [DataAPI.jl](https://github.com/JuliaData/DataAPI.jl));
       `Between(first, last)` selects the columns between `first` and `last`;
    * an `All` expression (see [DataAPI.jl](https://github.com/JuliaData/DataAPI.jl));
      `All()` selects all columns, equivalent to `:`;
```

File: docs/src/man/working_with_dataframes.md [hunk 3]
```diff
@@ -255,18 +255,49 @@ julia> df[!, Not(:x1)]
...
+
+julia> df[:, Cols(x -> startswith(x, "x"))] # keep columns whose name starts with "x"
+1×2 DataFrame
+ Row │ x1     x2
+     │ Int64  Int64
+─────┼──────────────
+   1 │     2      3
+```
+
+The following examples show a more complex use of `Cols` selector and move all
```
Comment (nalimilan) [inline docs/src/man/working_with_dataframes.md:297]:
```suggestion
The following examples show a more complex use of the `Cols` selector, which moves all
```

File: docs/src/man/working_with_dataframes.md [hunk 4]
```diff
@@ -255,18 +255,49 @@ julia> df[!, Not(:x1)]
...
+columns whose names match `r"x"` regular expression respectively to the front
+and to the end of a data frame:
```
Comment (nalimilan) [inline docs/src/man/working_with_dataframes.md:299]:
```suggestion
and to the end of the data frame:
```

File: src/other/index.jl (hunk 1 above)
Comment (nalimilan) [reply src/other/index.jl:234]: We really need page(s) about indexing and `select`/`transform`/... which would be more detailed than Getting Started, but less technical that the API reference.

It's a bit weird that we add a docstring to `Cols` that only makes sense for DataFrames. Maybe we should document this in DataAPI?

Comment (bkamins) [reply src/other/index.jl:234]: I have opened https://github.com/JuliaData/DataAPI.jl/pull/42 (and then we can remove the docstring here)
and https://github.com/JuliaData/DataFrames.jl/issues/2887 (to handle this in the future)

File: docs/src/lib/internals.md [hunk 5]
```diff
@@ -16,4 +16,5 @@ getmaxwidths
 ourshow
 ourstrwidth
 @spawn_for_chunks
+Cols
```
Comment (bkamins) [inline docs/src/lib/internals.md:19]:
```suggestion
```

File: src/other/index.jl (hunk 1 above)
Comment (bkamins) [inline src/other/index.jl:234]:
```suggestion
```

File: NEWS.md [hunk 6]
```diff
@@ -70,6 +70,8 @@
 * the `DataFrame` constructor when matrix is passed to it as a first
   argument now allows `copycols` keyword argument
   ([#2859](https://github.com/JuliaData/DataFrames.jl/pull/2859))
+* `Cols` now accepts a predicate accepting column names as strings.
+  ([#2880](https://github.com/JuliaData/DataFrames.jl/pull/2880))
```
Comment (bkamins) [inline NEWS.md:74]:
```suggestion
  ([#2881](https://github.com/JuliaData/DataFrames.jl/pull/2881))
```

**Comment by bkamins** [comment]
@nalimilan - this should be good to have a look at again.

**Review by nalimilan — APPROVED** [review]

**Comment by bkamins** [comment]
Thank you!

**Pull request merged**